### DISCUSSION TITLE
Bounds in optimization

### DISCUSSION THREAD

**Discussion by am-tc01** [idaholab/moose#26160] [Q&A Modules: General]
Hi,
I am working with optimization and realized that the parameters' values that the Optimize solver take do not strictly respect the `upper_bounds` and `lower_bounds`. Sometimes the parameters' value can go beyond the bounds as well. For example, in this [test](https://github.com/idaholab/moose/blob/next/modules/optimization/test/tests/optimizationreporter/nonlinear_material/main.i), the parameter value in `initial_condition` is smaller than the `lower_bounds`. It works even if I change the solver to TAOBLMVM. Is that supposed to be so?

Regards.

**Comment by GiudGiud** [comment]
Hello

Initial conditions are a special case, they are not coming from the optimization solve.
Do you get out of bounds values from the output of the solver?

**Reply by am-tc01** [reply]
But shouldn't it throw an error even if the initial value is out of bounds? The first iteration the solver takes, uses the initial value. What if it converges with the initial value?
I have made an example where I call it converged as soon as the value of objective function is smaller than the threshold. In such case, the output of the solver could be out of bounds. Let me see if I can make a reproduceable case.

**Reply by GiudGiud** [reply]
I dont see why we would impose that restriction. 
If the bounds are complicated to process, that would mean that the user has to do some significant work to just initialize the simulation well enough to meet the acceptance criterion

**Reply by am-tc01** [reply]
Hi,
So for this case, I get out of bound values from the output of the solver.
It seems that it converges for the given criteria but the output I get for the `heat_source` (301) is greater than the `upper_bounds` (300),
[moose_optimization_test.zip](https://github.com/idaholab/moose/files/13487265/moose_optimization_test.zip)

**Reply by GiudGiud** [reply]
@lynnmunday or @grmnptr want to look into it?
Maybe we re missing a check

**Reply by maxnezdyur** [reply]
```
[OptimizationReporter]
  type = OptimizationReporter
  parameter_names = 'heat_source'
  num_values = '1'
  initial_condition = '300'. <---- IC of 300 which is right at the upper bound
  lower_bounds = '0.1'
  upper_bounds = '300'
  measurement_points = '0.2 0.2 0
                        0.8 0.6 0
                        0.2 1.4 0
                        0.8 1.8 0'
  measurement_values = '1.98404 1.91076 1.56488 1.23863'
[]

tao_fd_delta = 1 <---- step size of 1 is used for finite differencing
gatol = 1E-8
grtol = 1E-8
gttol = 1E-8
fmin = 0.001

[Executioner]
  type = Optimize
  tao_solver = taoblmvm
  petsc_options_iname = '-tao_fd_gradient   -tao_fd_hessian     -tao_gatol    -tao_fmin    -tao_cmonitor  -tao_fd_delta'  <---- you are using finite differencing for the gradient
  petsc_options_value = ' true               true            	  ${gatol}	    ${fmin}		  	true      	   ${tao_fd_delta}'
  verbose = true
[]
```
You are using finite differencing with a step size of 1 (which I would make sure is not too large), and have an initial condition right at the upper bound. MOOSE does not control how TAO does FD, so I believe that's what you are getting a heat source of 301.

**Reply by am-tc01** [reply]
Thanks for looking into this! I just wanted to reproduce a case where the output of the solver is out of the bounds, hence the strange values. But if you change upper_bounds to say 299.5 you'll still get a value of heat_source (300.5) which is beyond the bounds. Of course it depends on FD_DELTA. But in some sense upper and lower bounds are not respected strictly.

**Reply by maxnezdyur** [reply]
> But if you change upper_bounds to say 299.5 you'll still get a value of heat_source (300.5) which is beyond the bounds.

Do you mean the converged value is 300.5 or that the first step is 300.5? There are other cases where the bounds are not strictly respected. IIRC if you use a "unit" line search when using the BQNLS method then the bound won't be respected during the line search. All the bounds are handled by TAO, so I would look at their documentation at possible conflicts with options used.

**Reply by am-tc01** [reply]
The converged value is 300.5

**Reply by maxnezdyur** [reply]
Can you paste a few lines of the monitor that gets printed out. I think it may be an outputting issue, where the last output is actually the FD gradient check aka 299.5 + 1 and the actual answer that you are converging on is 299.5. If you decrease the FD delta to 0.1, would  your final answer be 299.6?

**Reply by am-tc01** [reply]
This is the output from `OptimizationReporter` with FD delta=1:
```
grad_heat_source,heat_source,measurement_time,measurement_values,measurement_xcoord,measurement_ycoord,measurement_zcoord,misfit_values,simulation_values
0,300.5,0,1.98404,0.2,0.2,0,-0.0058508197670573,1.9781891802329
0,0,0,1.91076,0.8,0.6,0,-0.014190147372656,1.8965698526273
0,0,0,1.56488,0.2,1.4,0,-0.017321388655594,1.5475586113444
0,0,0,1.23863,0.8,1.8,0,-0.0093261938743641,1.2293038061256
```

and this is the output with FD delta=0.1:
```
grad_heat_source,heat_source,measurement_time,measurement_values,measurement_xcoord,measurement_ycoord,measurement_zcoord,misfit_values,simulation_values
0,299.6,0,1.98404,0.2,0.2,0,-0.0060129731557681,1.9780270268442
0,0,0,1.91076,0.8,0.6,0,-0.014584642680112,1.8961753573199
0,0,0,1.56488,0.2,1.4,0,-0.017803741485438,1.5470762585146
0,0,0,1.23863,0.8,1.8,0,-0.0095855652419727,1.229044434758
```
so yes, the final answer is 299.6

**Reply by lynnmunday** [reply]
Did @maxnezdyur answer this?  We could add a check in MOOSE that the initial_conditions fall within the range of upper and lower bounds.  We set the initial conditions on the MOOSE side and we never check that they are within the bounds.  I agree with everything Max said, bounds are handled by the line search algorithm in TAO.  If you use linesearch=unit, bounds will be ignored.  Finite differencing could also ignore the bounds.  Nelder Mead will also ignore bounds because it does not use linesearch.

**Reply by GiudGiud** [reply]
I would allow intial conditions out of bounds if it does not cause an error.
Ignoring initial conditioons is a problem though, you should fix that if it occurs

**Reply by lynnmunday** [reply]
@GiudGiud it won't ignore initial_conditions, those are always used if provided and if they are not provided, they default to zero.  I'm pretty sure, does someone have an example where that is not the case?

**Reply by am-tc01** [reply]
In my opinion, it's probably good to add a check that the initial conditions are not out of bounds. In BOBYQA, for instance, a runtime warning is thrown and the initial values are adjusted if they are out of bounds. Perhaps it's also good to add a check that `lower_bounds` are not greater than the `upper_bounds`. But I agree, the issue I have is probably coming from TAO and not from MOOSE. I can't formulate an analytical solution to my problem so I have to rely on FD. I realized that I can reach convergence faster if I use a large value of fd_delta, but that also means that solution could be significantly out of bounds, which is strange. I'll look more into TAO documentation to see if there is a conflict somewhere. 
For the example I posted above, it seems that the actual upper_bound is upper_bounds (that I define in moose) + Tao_fd_delta. So even if I define initial_condition (300) which is way above upper_bounds (250), it gives me a solution within upper_bounds + tao_fd_delta (250+0.5) and not initial_condition + tao_fd_delta.

**Reply by lynnmunday** [reply]
Initial conditions and bounds are mostly handled in TAO and TAO will usually project the initial conditions onto the bounds if it is initially outside the bounds with normal settings.  Finite differencing, line search options, gradient free options might modify this but that is documented and checked by TAO.  I'm going to leave all of the checking to TAO since that is where these options are implemented and documented.  We also wouldn't be able to catch the finite difference problem you're running into.
